add transpose method to _Audio

artpelling opened the pull request:
This PR adds a `transpose` method to all `Signals`, very similar to `np.ndarray.transpose`.

f-brinkmann commented on pyfar/classes/audio.py:
```diff
@@ -108,6 +108,28 @@ def reshape(self, newshape):
 
         return reshaped
 
+    def transpose(self, *axes):
```
I would suggest `axes=None` as default as in numpy

artpelling replied:
I did that but then realized that this does not reproduce the behaviour of numpy. `np.transpose` uses `axes=None` per default, however, `np.ndarray.transpose` uses `*axes` which allows passing of tuples and multiple ints.

I have reverted these changes, left `axes` empty per default and also accept `None`.

f-brinkmann commented on pyfar/classes/audio.py:
```diff
@@ -108,6 +108,28 @@ def reshape(self, newshape):
         return reshaped
 
+    def transpose(self, *axes):
+        """Return a transposed copy of the audio object.
```
Maybe even more explicit
```suggestion
        """Transpose time/frequency data and return copy of the audio object.
```

f-brinkmann commented on pyfar/classes/audio.py:
```diff
@@ -108,6 +108,28 @@ def reshape(self, newshape):
+        axes : empty, iterable of ints, or n ints
+            leaving the argument empty will reverse the order of `self.caxes`.
+            iterable of ints: `i` in the `j`-th place of the interable means
+            that `self._data`'s `i`-th axis becomes the transposed datas `j`-th
```
I would avoid mentioning `_data` in the docstring because we usually hide this from the user. Maybe soemthing like '... means that the `i`-ths axis of the time/frequency data becomes ...'

f-brinkmann commented on pyfar/classes/audio.py:
```diff
@@ -108,6 +108,28 @@ def reshape(self, newshape):
+            n ints: same as 'iterable of ints'.
+        """
+        axes = tuple(range(len(self.cshape)))[::-1] if len(axes) == 0 else tuple(axes)
+        assert all([isinstance(ax, int) for ax in axes]), "axes must be defined as integers"
```
Nice, did not know the shorthand for throwing a specific error message

f-brinkmann commented on pyfar/classes/audio.py:
```diff
@@ -108,6 +108,28 @@ def reshape(self, newshape):
+        axes = tuple(range(len(self.cshape)))[::-1] if len(axes) == 0 else tuple(axes)
+        assert all([isinstance(ax, int) for ax in axes]), "axes must be defined as integers"
+        assert len(axes) == len(set(axes)), "axes have to be unique"
+        assert len(axes) == max(axes) + 1 == len(self.cshape), "axes have to match caxes"
```
I think this assert will fail if specifying negative axes, e.g., `axes = (0, -1)`

artpelling replied:
Ah thanks. I totally missed that negative axes are admissible in `np.transpose`.

f-brinkmann reviewed (changes requested):
Thanks for the PR. Only a few comments on the docstring and some gneral points to consider/discuss

- If we specify axes for audio objects we have the concept of ignoring the last axis holding the time frequency bins. This means, that axis `-1` in an audio object refers to axis `-2` of the underlying numpy array `_data`. You can account for this by `axes = [a-1 if a<0 else a for a in axes]` and add a description for this in the docstring.
- I noted that the test are failing. Might be because you branched main. Can you branch develop instead?
- Once the code is approved we should add tests to **tests/test_audio_signal.py**, **tests/test_audio_time_data.py**, and **tests/test_audio_frequency_data.py**

artpelling commented:
I've implemented your suggestions. Out of lazyness, I removed all assertions and simply transpose a mock array with identical `ndim` before deep copying instead of reverse engineering all rules. This will throw numpy's exceptions. Let me know if you like it or not.

> * Once the code is approved we should add tests to **tests/test_audio_signal.py**, **tests/test_audio_time_data.py**, and **tests/test_audio_frequency_data.py**

Will do that once the code is agreed on.

ahms5 reviewed (commented):
Thanks for the PR, very elegant, looks fine to me. for me just tests are remaining. flake8 should be fixed with #483

artpelling commented:
@f-brinkmann This might be ready now.

f-brinkmann reviewed (changes requested):
Thanks, looks very good already. I only had minor comments on the docstring and testing left. Can you also add a brief test for the shorthand signal.T?

artpelling commented on pyfar/classes/audio.py:
```diff
@@ -95,11 +95,40 @@ def reshape(self, newshape):
+        if axes is None or len(axes) == 0:
+            axes = tuple(range(len(self.cshape)))[::-1]
+        else:
+            tuple([a-1 if a < 0 else a for a in axes])
```
I think it is fine, this will lead to errors in both 
```python
np.empty(np.ones(len(self.cshape), dtype=int)).transpose(axes)
``` 
as well as
```python
transposed._data = transposed._data.transpose(*axes, len(self.cshape))
```

artpelling commented on tests/test_audio_frequency_data.py:
```diff
@@ -104,10 +104,22 @@ def test_reshape_exceptions():
         data_out = data_in.reshape((3, 4))
 
 
+def test_transpose():
```
I feel like a lot of the tests could be reduced that way. Essentially the files `test_audio_signal.py`, `test_audio_frequency_data.py` and `test_audio_time_data.py`could be merged.

f-brinkmann replied:
I never checked - you could make an issue so we don't forget to discuss it on the weekly.

artpelling replied:
Done. #516

f-brinkmann reviewed (dismissed):
There seems to be an error in the tests now, but otherwise approved. Thanks for the effort.
